drivers/fpga/dfl-pci-sva.c

```diff
@@ -0,0 +1,324 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Manage PASID and IOMMU binding for FPGA ports supporting shared
+ * virtual addressing.
+ *
+ * Copyright (C) 2023 Intel Corporation, Inc.
+ *
+ * Authors:
+ *   Michael Adler <michael.adler@intel.com>
+ */
+
+/*
+ * Expose a device interface for binding a PASID to FPGA PCIe ports.
+ * PCIe devices with a DFL device ID are tracked independent of their
+ * current driver, making it possible to configure IOMMU SVA with
+ * either dfl-pci or vfio-pci drivers. The module uses bus notifier
+ * hooks in order to catch device removal when vfio-pci is bound.
+ *
+ * Device files named /dev/dfl-pci-sva/<pci address> are managed
+ * in this module (e.g. /dev/dfl-pci-sva/0000:ab:00.3). The
+ * DFL_PCI_SVA_BIND_DEV ioctl ensures a PASID is allocated, binds
+ * it to the device and returns the PASID. The PASID remains bound
+ * until the file is closed or DFL_PCI_SVA_UNBIND_DEV.
+ */
+
+#include <linux/pci.h>
+#include <linux/types.h>
+#include <linux/kernel.h>
+#include <linux/notifier.h>
+#include <linux/iommu.h>
+#include <linux/miscdevice.h>
+#include <linux/fpga-dfl.h>
+
+#include "dfl.h"
+
+static struct notifier_block sva_nb;
+
+static DEFINE_MUTEX(dfl_dev_list_mutex);
+static LIST_HEAD(dfl_dev_list);
+
+struct dfl_sva_dev {
+	struct pci_dev *pdev;		  /* PCIe device to bind */
+	struct miscdevice mdev;		  /* /dev/dfl-pci-sva/<addr> device */
+	char mdev_name[64];		  /* dfl-pci-sva!<addr> */
+	struct file_operations mdev_fops; /* dfl-pci-sva file ops */
```

Comment:
> Could it be declared and initialized as a single static instance and then be shared by all mdev devices?

No, it can't. I'm using mdev_fops to find the parent struct with container_of for all file handles. Is there a better way to do this?